### DISCUSSION TITLE
Match All Criteria and Faceted Filtering With Front End API

### DISCUSSION THREAD

**Discussion by gmcleod3308** [swellstores/community#108] [Help] [answered]
Hey everyone, sorry for the world's longest title, I just didn't know how to word the question.

We're building a faceted search using the front end api.

I have a query that looks like this

`  params = {
          active: true,
          page: "1",
          limit: "24",
          expand: [
              "categories"
          ],
          $filters: {
              category: {$in: ["grade-mid-grade"]},
              length:["4.5"]
          }
 }`
 
 `const response = await swell.products.list(params);`
 
 This query is returning two records, one of the records matches the category and the length, the other record only. matches the length.
 
 To me this is incorrect.  I would expect the query to only return 1 result since the second record didn't match all criteria.

So my question is this the intended behavior or is this a bug?  and if this is the intended behavior is there a way to create a $filter so that the results only includes records that match all criteria? 

Thank you,

**Comment by swellmike** [comment, marked as answer]
I don't think the `category` object filter will work as you intend.  If you know the category ID, you could instead filter by `category_index`, which contains a denormalized mapping of related category IDs:

```
"category_index.id": "<category ID>"
```

**Reply by ericingram** [reply]
To add to that: the `category: { $in }` filter essentially does an `$or` query. If you want to control the logical flow, you'll need to make a query along those lines using something like `$and: [ { 'category_index.id': ... }, ... ]`

**Reply by gmcleod3308** [reply]
Thanks guys, so basically there's no way to do this with just the slug?  I'll need to track the id's of the categories as well and build a where clause.  So that brings up another question.  

The chat support referred me to this thread here https://github.com/orgs/swellstores/discussions/63. which seems to be the same issue more or less.  But the solution is using the backend API  syntax. so far I've been using the front end API, so can we use both interchangeably? Our app is a nuxtjs app with vue so most of the queries are happening in Vue components which is why I thought the Front End API made the most sense. 

Thanks again for the help.  I'll see if I can build a where with $and and category id's and report back.

**Reply by ericingram** [reply]
That's right, it doesn't work with slugs yet, but agree in general it would be great if any ID values in queries could be substituted with secondary keys.

Besides that we'll consider enhancing `$filters` to support more complex logic in the future.

**Reply by swellmike** [reply]
> But the solution is using the backend API syntax. so far I've been using the front end API, so can we use both interchangeably?

The overall structure of the API requests (filters, projections, pagination, etc.) are the same between frontend and backend APIs.  The main difference between the frontend and backend APIs is permissions.  The frontend API offers only a subset of API endpoints, and within that subset, it is further restricted via permissions (some fields are read-only, some are not available, some require an authenticated session).

A couple thoughts on your use case:
1. for a storefront (frontend), you'll want to use the Frontend API. Otherwise you could risk giving end-users (customers) access to endpoints that could affect orders, products, etc.
2. it seems that `products.category_index` might not be readily available in the Frontend API, so you may have to customize the storefront permissions on your public key.

**Reply by gmcleod3308** [reply]
Hey all,

 I just wanted to give you all an update on this and to say thank you.  The query we ended up going with based on your suggestions looks like this
 
```
const query = {
    "active": true,
    "page": 1,
    "limit": "12",
    "search": "scissors",
    "expand": [
        "categories"
    ],
    "$and": [
        {
            "category_index.id": {
                "$in": [
                    "638d2c802c9786001274046d",
                    "638d2c992c9786001274046f",
                    "638d2bbfe6cb5f0013c18976"
                ]
            }
        },
        {
            "attributes.length": {
                "$in": [
                    "4.5",
                    "5"
                ]
            }
        },
        {
            "attributes.curvature": {
                "$in": [
                    "Sharp/Sharp",
                    "Blunt/Blunt"
                ]
            }
        }
    ]
}
```


It basically says show me all the scissors that are in these 4 categories that also have a length attribute of 4.5 or 5 inches long and a curvature attribute of Sharp/Sharp or Blunt/Blunt and it's working as expected. 

So again thank you all for taking the time to point us in the right direction.
